clients/drcachesim/tools/schedule_stats.cpp

```diff
@@ -0,0 +1,295 @@
+{
+    if (shard_type != SHARD_BY_CORE)
+        return "Only core-sharded operation is supported";
+    return "";
+}
+
+bool
+schedule_stats_t::process_memref(const memref_t &memref)
+{
+    error_string_ = "Only core-sharded operation is supported.";
+    return false;
+}
+
+bool
+schedule_stats_t::parallel_shard_supported()
+{
+    return true;
+}
+
+void *
+schedule_stats_t::parallel_shard_init_stream(int shard_index, void *worker_data,
+                                             memtrace_stream_t *stream)
+{
+    auto per_shard = new per_shard_t;
+    std::lock_guard<std::mutex> guard(shard_map_mutex_);
+    per_shard->stream = stream;
+    per_shard->core = stream->get_output_cpuid();
+    shard_map_[shard_index] = per_shard;
+    return reinterpret_cast<void *>(per_shard);
+}
+
+bool
+schedule_stats_t::parallel_shard_exit(void *shard_data)
+{
+    // Nothing (we read the shard data in print_results).
+    return true;
+}
+
+std::string
+schedule_stats_t::parallel_shard_error(void *shard_data)
+{
+    per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
+    return per_shard->error;
+}
+
+bool
+schedule_stats_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
+{
+    per_shard_t *shard = reinterpret_cast<per_shard_t *>(shard_data);
+    if (knob_verbose_ >= 4) {
+        std::ostringstream line;
+        line << "Core #" << std::setw(2) << shard->core << " @" << std::setw(9)
+             << shard->stream->get_record_ordinal() << " refs, " << std::setw(9)
+             << shard->stream->get_instruction_ordinal() << " instrs: input "
+             << std::setw(4) << shard->stream->get_input_id() << " @" << std::setw(9)
+             << shard->stream->get_input_interface()->get_record_ordinal() << " refs, "
+             << std::setw(9)
+             << shard->stream->get_input_interface()->get_instruction_ordinal()
+             << " instrs: " << std::setw(16) << trace_type_names[memref.marker.type];
+        if (type_is_instr(memref.instr.type))
+            line << " pc=" << std::hex << memref.instr.addr << std::dec;
+        else if (memref.marker.type == TRACE_TYPE_MARKER) {
+            line << " " << memref.marker.marker_type
+                 << " val=" << memref.marker.marker_value;
+        }
+        line << "\n";
+        std::cerr << line.str();
+    }
+    if (memref.marker.type == TRACE_TYPE_MARKER &&
+        memref.marker.marker_type == TRACE_MARKER_TYPE_CORE_WAIT) {
+        ++shard->counters.waits;
+        if (!shard->prev_was_wait) {
+            shard->thread_sequence += '-';
+            shard->cur_segment_instrs = 0;
+            shard->prev_was_wait = true;
+        } else {
+            ++shard->cur_segment_instrs;
+            if (shard->cur_segment_instrs == knob_print_every_) {
+                shard->thread_sequence += '-';
+            }
+        }
+        return true;
+    } else
+        shard->prev_was_wait = false;
```

Comment:
Ideally you'd set this at the end, just in case some other logic below wants to know whether prev_was_wait.

Reply:
Moved to the end.